src/draw/ElkLayout.js

```diff
@@ -0,0 +1,344 @@
+        'elk.layered.interactiveReferencePoint': this.elkParams.interactiveReferencePoint,
+      } : {}),
+      'elk.debugMode': 'true',
+
+      'elk.direction': this.elkParams.direction,
+    };
+
+    // we prepare the input in the format expected by ELK
+    const graph = {
+      id: parentNode.raw ? parentNode.raw.id : 'root', layoutOptions, children: [], edges: [],
+    };
+
+    graph.children = parentNode.children
+    // convert to the format expected by ELK
+      .map((node) => ({
+        id: node.raw.id,
+        width: node.raw.drawOption.width,
+        height: node.raw.drawOption.height,
+        x: node.raw.drawOption.x,
+        y: node.raw.drawOption.y,
+        layoutOptions: { ...layoutOptions },
+      }));
+
+    // we get the links that have source and target in the children of `node`
+    // TODO : what about cross-level links ? (ie. from a depth to another)
+    const localLinks = this.__getLinksForChildren(nodes, links, parentNode);
+
+    // convert the edges to the format expected by ELK
+    graph.edges = localLinks.map((link) => ({
+      id: `${link[0]}__${link[1]}`, sources: [link[0]], targets: [link[1]],
+    }));
+
+    // finally calling ELK
+    return ElkLayout.__elk.layout(graph);
+  }
+
+  /**
+   * Write a generated layout at a single level of hierarchy to the graph.
+   * @param {ElkNode} elkNode - previously generated layout with a single level of hierarchy.
+   * @private
+   */
+  __writeSingleDepthLayout(elkNode) {
+    const nodes = new Map(elkNode.children
+      .map((node) => [node.id, {
+        x: node.x, y: node.y,
+      }]));
+
+    this.pluginData.components.forEach((component) => {
+      if (nodes.has(component.id)) {
+        const { x, y } = nodes.get(component.id);
+
+        component.drawOption.x = x;
+        component.drawOption.y = y;
+      }
+    });
+  }
+
+  /**
+   * Compute a node's depth (ie. distance to the root).
+   * @param {NodeData} node - a node.
+   * @returns {number} depth.
+   * @private
+   */
+  __computeNodeDepth(node) {
+    if (!node.parent) {
+      // this is the root
+      return 0;
+    }
+
+    return this.__computeNodeDepth(node.parent) + 1;
+  }
+
+  /**
+   * Keep only the links that are contained within {@link nodes} children.
+   * @param {Map<string,NodeData>} nodes - all the nodes.
+   * @param {ComponentLink[]} allLinks - all the links.
+   * @param {NodeData} parentNode - node whose we want the inner links.
+   * @returns {[string,string][]} inner links represented as pairs of nodes ids.
+   * @private
+   */
+  __getLinksForChildren(nodes, allLinks, parentNode) {
+    // ids of parentNode's children
+    const keptIds = new Set(parentNode.children
+      .map((node) => node.raw?.id));
+    // depth at which we operate
+    const depth = parentNode.depth + 1;
+
+    return allLinks
+    // only links with valid source/target
+      .filter((link) => nodes.get(link.source) !== undefined
```

Comment:
remove this, can't have link without source or target.

Reply:
done